### DISCUSSION TITLE
Announcing Magmide Month!

### DISCUSSION THREAD

**Discussion by blainehansen** [blainehansen/magmide#28] [Announcements]
Hello everyone! I'm excited to let you know I'll be taking a month of sabbatical from my normal full-time job starting today! I've decided to use this time to make some real progress on Magmide, and I'm announcing that intention mostly to put myself on the hook. I want to use this time wisely! I'll finally have enough spare time and energy to push hard on the next important milestones of this project. I've decided to write weekly updates as well, so watch for those.

In this month I hope to:

- Figure out a basic paradigm for [stored program](https://en.wikipedia.org/wiki/Von_Neumann_architecture) operational semantics, which isn't usually how verification is carried out ([as discussed in my previous announcement](https://github.com/magmide/magmide/discussions/25)) (*about a few days of work*).
- Figure out using a toy operational semantics how [trackable effects](https://github.com/magmide/magmide#gradually-verifiable) could precisely work, culminating in just a blog post discussing the concept in general (*about a week of work*).
- Actually instantiate Iris over a stored program operational semantics and define a custom weakest precondition (*hopefully by the end of the month*). [I also discussed this in my previous announcement](https://github.com/magmide/magmide/discussions/25), and in snatches of time since then I've been able to better piece together how Iris uses the "magic wand" separation logic operator to define weakest preconditions and therefore operational semantics in general (section 6 of the [Iris paper](https://people.mpi-sws.org/~dreyer/papers/iris-ground-up/paper.pdf) is the important one to understand).

If I find myself stuck on something this month, I've resolved to ask for help from experts and the community more readily than my instincts would usually prompt me to. I need to get over my predilection to just grind on something by myself for way too long before seeking help, so watch for that as well!

At the end of the month if things have gone well I'll likely be making a general call for collaborators, especially those already knowledgeable about Iris, Coq, or LLVM. I'm not terribly motivated by writing academic papers (everything I ever do will primarily be exposed to the world in blog posts and project documentation), but there could be a few sufficiently novel concepts in the path of this project to make it worth working with some academics to publish something somewhere. Perhaps that would help draw in other helpful experts.

Thank you all very much! I'm excited for the coming month!

**Comment by blainehansen** [comment]
Hello everyone! This is my first weekly update, and I'm happy to say my progress over the past few days makes me hopeful!

In my announcement on Tuesday, I sketched out a plan where I first figured out some basic semantics, then do some design writing about trackable effects, and then tackle the Iris work. I quickly realized after toying with a semantics proposition that I wouldn't actually be able to get much of *interest* done before I got my hands dirty and actually started using Iris. Almost anything I would have done before that would have ended up being thrown away.

In the past I've never been disappointed when I choose to do the *hardest and most unknown thing first*, so I decided my first goal would be to instantiate Iris enough to start truly using and experimenting with it! After a few brain-melting days trying to truly understand some of the finer conceptual points and most importantly spelunking the actual Iris codebase to try to grok how the theoretical concepts were achieved in real code, I can finally say I understand it all well *enough* to know what the path forward looks like. I still don't have a ton of code written, but I'm finally at the point where I can actually start asking *design* questions about how I'll use Iris, rather than merely staring in terrified mystification at the papers or codebase.

In general the hard questions in front of me all somehow relate to the ways in which the Magmide project has different goals and structures than existing academic projects:

- I'm starting by modeling "stored program" machines, which are very different than "functional expression" languages that it seems the academic crowd almost exclusively works in. I think this is a very important thing to do for two reasons (apart from being interesting to me ha):
	- All higher level languages get their semantics from the lowest level assembly architectures, and it's important in the long term to be able to verify extremely important things like register calling conventions or self-mutating programs "from the bottom up".
	- If this project is actually going to be useful to someone very quickly, it might make sense to figure out how to provide verification tools to people writing *very* low level code, such as [P4](https://en.wikipedia.org/wiki/P4_(programming_language)) or other commonly deployed embedded architectures such as [6502](https://en.wikipedia.org/wiki/MOS_Technology_6502). I haven't looked around to understand *exactly* which crowd might be open to this, but it seems possible.
- I'm working towards trackable effects, whose entire point is to allow "progressive verification", meaning a program can do *anything* that's physically possible in the machine, but if a program performs some action without providing a proof that action abides by some invariant then that code will be "marked" with a trackable effect that "bubbles" to consumers when they call that code. Consumers can either accept that "transitive marking" or somehow remove the effect with sandboxing or by transforming the marked code. This is very different than the specifications academics are creating where something like accessing unallocated memory is simply ill-typed and not allowed.

The main consequence of those differences is that I probably have to use slightly customized Resource Algebras to model finite register banks and memory spaces, rather than the infinite and explicitly allocatable heaps academics are reasoning about. I've been figuring out exactly what I should use to avoid using some off-the-shelf concepts that might end up getting in my way.

But despite having some tricky questions in front of me, I'm finally actually confident I can resolve them and get somewhere!

P.S. I'll likely do an extended tutorial of Iris for non-academics after this month is over. I'm waiting until the end because it will be pretty easy for me to write a tutorial like that on my off-hours once I go back to my full-time job, but during this month I want to save my deep work energy for this extremely demanding Iris stuff.

Thank you all! See you next week!

**Comment by blainehansen** [comment]
Hello everyone!

Quick update this week, I don't have time to ramble as much as I want to ha.

This previous week was frustrating, had some annoying personal logistical stuff come up that knocked out a couple days.
Realized I needed to focus on building an [LLVM analogue](https://github.com/magmide/magmide/blob/main/posts/design-of-magmide.md), not a "pure" assembly language, so redirected my efforts. Had to read a paper or two to realize that and then did yet more reading to try to better understand Iris's "Lob induction", which actually allows a form of [*co*induction](http://adam.chlipala.net/cpdt/html/Coinductive.html).
Feeling a bit mired in indecision, but going to keep plugging!

We'll see what this week brings.

**Reply by zicklag** [reply]
🏃 Just want to say that I love the vision of this project, and good luck with your efforts and research! 🙂 🚀

**Reply by blainehansen** [reply]
Thank you I appreciate that!

**Comment by blainehansen** [comment]
Hey everyone! Sorry about the long gap between updates, but I spent this previous weekend and the beginning of this week scrambling to rework a trip to Paris I had planned a few months ago (my travel companion didn't want to find out how historic the protests and strikes were going to be ha). I'm in Italy right now, and have been constantly going going going for a few days.

Honestly I'm a little frustrated with the progress I've made on Magmide in the last few weeks. The first week was great, I really did push through some understanding barriers and got a lot more mentally equipped. But after that I let myself get bogged down reading papers and the Iris codebase, and haven't been following through with my resolution to stay practical and scrappy by focusing on writing and checking code. Part of it is that I burned myself out a bit that first week, but mostly I think it's just been the general insecurity of pursuing a project where I have very little official support/mentorship, and where the work draws very little on my existing professional experience. As I've said in the past, I wish I could pursue this project in the context of a PhD, but that's simply not in the cards.

I'll be back from Europe this coming Wednesday, so I'll have some time the days after to do a bit of work on Magmide, and then the Monday after is when my sabbatical ends! I didn't appreciate how burned out I was entering this sabbatical, so going forward I'm going to double down on pursuing a sustainable schedule of [slow productivity](https://calnewport.com/on-slow-productivity-and-the-anti-busyness-revolution/), consistently chipping away every day/week rather than trying to set aside huge amounts of time. As heroic as it can feel to burn the midnight oil and make flashes of progress, I'm finding this project simply too big and uncertain to make that a wise course of action.

Thank you all! I'm sorry this ended up being a bit disappointing, but I guess that's one of the struggles of independent research.

P.S. I'm considering starting a weekly newsletter just about my personal thoughts and ramblings. I would only do it to give myself a way to work in public and commit to consistency, so it wouldn't only be about Magmide, since sometimes I have other things it makes sense to jump to. Let me know if that seems extremely exciting to you, I'm still on the fence.

**Reply by zicklag** [reply]
I think starting a newslwetter might be a good idea. So far you seem to be intersted in enough thought provoking concepts and interesting ideas. I'm guessing I'd probably read it, from what I've seen from you so far.

I do a _lot_ of random experiments and trying stuff out or learning about new things, and I've needed to get better at writing about it, so recently I just started my own blog. 

I've realized that it can be pretty enlightenting to read about somebody else's experience, even if it's the first time they've done something. If you've never done that thing before either, it can be extra useful sometimes, because they are comming at the problem with something like your perpsective.

I started my own [blog repo](https://github.com/zicklag/blog) on GitHub with automatic deployments to GitHub pages if you want to use it as a template. ( If you don't already have something setup or in mind. )

It's built on Zola, so you could grab any of a number of [themes](https://www.getzola.org/themes/).

**Reply by zicklag** [reply]
**PS:** The technology I've been looking into recently is [HVM](https://github.com/HigherOrderCO/HVM) which might have something interesting to offer your study for magmide. It's not got the same vision or goal, but some of the computation theory of Interaction Nets/Combinators might be useful or else just interesting.

---

**PPS:** No problem about the "disappointment". 😄 I know a little bit about how it goes just trying to figure stuff out and make things happen when you're not "qualified" for it. It's exciting, but also difficult, and sometimes seems pointless. I figure, if I'm still learning, I never know how what I do might be useful, even if it doesn't pan out how I thought it would.

**Comment by charlielidbury** [comment]
I am a computing student a few weeks away from embarking on my masters. For my masters, I intent on doing a proof of concept for a compilation technique for combining mutability and dependent types for a language I indend to make, [Ochre](https://charlielidbury.com/ideas/ochre).
> there could be a few sufficiently novel concepts in the path of this project to make it worth working with some academics to publish something somewhere

I am way behind you knowledge wise, but I intend on following a similar path as you and I'm looking into doing a PhD so I might be the academic to publish these novel concepts.

I've reached out on Discord & LinkedIn in case you would be up for talking, it might turn out parts of my project could be co-authored with you, as there seems to be overlap. Also, if this ends up being publishable I would be able to do it full time as part of a PhD.

**Reply by blainehansen** [reply]
Happy to hear more as you progress through your degrees and projects :) I'll continue to give updates as I make progress, so if things come up I'll keep you in mind!
